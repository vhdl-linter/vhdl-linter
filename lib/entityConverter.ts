import { Position } from 'vscode-languageserver';
import { findObjectFromPosition } from './languageFeatures/findObjects';
import { OArchitecture, OEntity, OGeneric, OPort } from './parser/objects';
import { ISettings } from './settingsGenerated';
import { VhdlLinter } from './vhdlLinter';


function longestInArray(array: OPort[] | OGeneric[]) {
    let longest = 0;
    for (const object of array) {
        if (object.lexerToken.text.length > longest) {
            longest = object.lexerToken.text.length;
        }
    }
    return longest;
}
export function instanceTemplate(entity: OEntity, settings: ISettings) {
    let text = `${settings.style.instantiationLabelPrefix}${entity.lexerToken.text}${settings.style.instantiationLabelSuffix} : entity work.${entity.lexerToken.text}`;
    const indentString = '  ';
    if (entity.generics.length > 0) {
        text += `\ngeneric map (\n`;
        const longest = longestInArray(entity.generics);
        for (const generic of entity.generics) {
            const name = generic.lexerToken.text.padEnd(longest, ' ');
            text += `${indentString}${name} => ${settings.style.genericPrefix}${generic.lexerToken.text}${settings.style.genericSuffix},\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n)`;
    }

    if (entity.ports.length > 0) {
        text += `\nport map (\n`;
        const longest = longestInArray(entity.ports);
        for (const port of entity.ports) {
            const name = port.lexerToken.text.padEnd(longest, ' ');
            text += `${indentString}${name} => ${settings.style.signalPrefix}${port.lexerToken.text}${settings.style.signalSuffix},\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n)`;
    }

    text += `;\n`;
    return text;
}
export function componentTemplate(entity: OEntity) {
    let text = `component ${entity.lexerToken.text} is`;
    const indentString = '  ';
    if (entity.generics.length > 0) {
        text += `\ngeneric (\n`;
        for (const generic of entity.generics) {
            text += `${indentString}${generic.range.getText()};\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n);`;
    }

    if (entity.ports.length > 0) {
        text += `\nport (\n`;
        for (const port of entity.ports) {
            text += `${indentString}${port.range.getText()};\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n);`;
    }

    text += `\nend component;\n`;
    return text;
}

export function sysVerilogTemplate(entity: OEntity, settings: ISettings) {
    let text = entity.lexerToken.text;
    const indentString = '  ';
    if (entity.generics.length > 0) {
        text += ` #(\n`;
        const longest = longestInArray(entity.generics);
        for (const generic of entity.generics) {
            const name = generic.lexerToken.text.padEnd(longest, ' ');
            text += `${indentString}.${name}(${settings.style.genericPrefix}${generic.lexerToken.text}${settings.style.genericSuffix}),\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n)`;
    }

    if (entity.ports.length > 0) {
        text += ` ${settings.style.instantiationLabelPrefix}${entity.lexerToken.text}${settings.style.instantiationLabelSuffix} (\n`;
        const longest = longestInArray(entity.ports);
        for (const port of entity.ports) {
            const name = port.lexerToken.text.padEnd(longest, ' ');
            text += `${indentString}.${name}(${settings.style.signalPrefix}${port.lexerToken.text}${settings.style.signalSuffix}),\n`;
        }
        // Strip the final comma
        text = text.slice(0, -2);
        text += `\n)`;
    }

    text += `;\n`;
    return text;
}
export function signalsTemplate(entity: OEntity, settings: ISettings) {
    let text = '';
    if (entity.ports.length > 0) {
        const longest = longestInArray(entity.ports);
        for (const port of entity.ports) {
            const name = port.lexerToken.text.padEnd(longest, ' ');
            let typeText: string | undefined;
            if (port.subtypeIndication.typeNames.length > 0) {
                typeText = port.subtypeIndication.typeNames[0]!.range.copyWithNewEnd(port.range.end).getText();
            }
            text += `signal ${settings.style.signalPrefix}${name}${settings.style.signalSuffix} : ${typeText ?? 'unknown_type'};\n`;
        }
    }
    return text;
}
export type converterTypes = 'instance' | 'signals' | 'sysverilog' | 'component';
export function entityConverter(vhdlLinter: VhdlLinter, type: converterTypes, settings: ISettings, position?: Position) {
    let entity;
    if (position === undefined) {
        entity = vhdlLinter.file.entities[0];
    } else {
        const rootElement = findObjectFromPosition(vhdlLinter, position)[0]?.getRootElement();
        if (rootElement instanceof OEntity) {
            entity = rootElement;
        } else if (rootElement instanceof(OArchitecture) && rootElement.correspondingEntity) {
            entity = rootElement.correspondingEntity;
        } else {
            // Fallback to first entity
            entity = vhdlLinter.file.entities[0];
        }
    }
    if (!entity) {
        return;
    }
    if (type === 'instance') {
        return instanceTemplate(entity, settings);
    } else if (type === 'signals') {
        return signalsTemplate(entity, settings);
    } else if (type === 'sysverilog') {
        return sysVerilogTemplate(entity, settings);
    } else if (type === 'component') {
        return componentTemplate(entity);
    }
}