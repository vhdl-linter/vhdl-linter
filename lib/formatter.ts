class Formatter {
    private pos: number = 0;
    constructor(public text: string) {

    }
    gotoChar(pos: number) {
        if (pos === -1) {
            this.pos = this.text.length - 1;
        } else {
            this.pos = pos;
        }
    }
    vhdlBeautifyRegion() {
        this.vhdlFixupWhitespaceRegion();
        // this.vhdlFixStatementRegion();
        // this.vhdlIndentRegion();
        // this.vhdlAlignRegion();
        // this.vhdlFixCaseRegion();
        // this.vhdlRemoveTrailingSpacesRegion();
        // Fix Tabs
        return this.text;
    }
    vhdlFixStatementRegion() {
        throw new Error("Method not implemented.");
    }
    vhdlIndentRegion() {
        throw new Error("Method not implemented.");
    }
    vhdlAlignRegion() {
        throw new Error("Method not implemented.");
    }
    vhdlFixCaseRegion() {
        throw new Error("Method not implemented.");
    }
    vhdlRemoveTrailingSpacesRegion() {
        throw new Error("Method not implemented.");
    }
    match: RegExpExecArray;
    matchStart: number;
    matchEnd: number;
    reSearchForward(re: RegExp | string) {
        if (typeof re === 'string') {
            re = new RegExp(re);
        }
        const match = re.exec(this.text.substring(this.pos));
        if (match === null) {
            return false;
        }
        this.matchStart = this.pos + match.index;
        this.matchEnd = this.matchStart + match[0].length;
        return true;
    }
    replaceMatch(replacement: string, subgroup?: number) {
        if (subgroup === undefined) {
            this.text = this.text.substring(0, this.matchStart) + replacement + this.text.substring(this.matchEnd);
        } else {
            let newReplacement = '';
            for (let i = 0; i < this.match.length; i++) {
                if (i === subgroup) {
                    newReplacement += replacement;
                } else {
                    newReplacement += this.match[i];
                }
            }
            this.text = this.text.substring(0, this.matchStart) + newReplacement + this.text.substring(this.matchEnd);
        }
        // TODO: Correct Cursor?
    }
    vhdlFixupWhitespaceRegion() {
        //         (goto - char end)
        //         (setq end(point - marker))


        //         ;; have no space before and one space after`,' and ';'
        //     (goto-char beg)
        this.gotoChar(0);
        //     (while (re-search-forward "\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\)\\|\\(\\s-*\\([,;]\\)\\)" end t)
        //       (if (match-string 1)
        // 	  (goto-char (match-end 1))
        // 	(replace-match "\\3 " nil nil nil 2)))
        while (this.reSearchForward("\\(--.*\\n\\|\"[^\"\\n]*[\"\\n]\\|'.'\\|\\\\[^\\n]*[\\n]\\)\\|\\(\\s-*\\([,;]\\)\\)")) {
            this.gotoChar(this.matchEnd);
            this.replaceMatch(this.match[3], 2);
        }
        //     ;; have no space after `('
        //             (goto - char beg)
        this.gotoChar(0);
        //             (while (re - search - forward "\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\)\\|\\((\\)\\s-+" end t)
        //         (if (match - string 1)
        //         (goto - char(match - end 1))
        //             (replace - match "\\2")))
        // while (this.reSearchForward("\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\)\\|\\((\\)\\s-+")) {
        //     this.gotoChar(this.matchEnd);
        //     this.replaceMatch(this.match[3], 2);
        // }



        //         ;; have no space before `)'
        //     (goto-char beg)
        //     (while (re-search-forward "\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\|^\\s-+\\)\\|\\s-+\\()\\)" end t)
        //       (if (match-string 1)
        // 	  (goto-char (match-end 1))
        // 	(replace-match "\\2")))
        //     ;; surround operator symbols by one space
        //     (goto-char beg)
        //     (while (re-search-forward "\\(--.*\n\\|\"[^\"\n]*[\"\n]\\|'.'\\|\\\\[^\\\n]*[\\\n]\\)\\|\\(\\([^/:<>=\n]\\)\\(:\\|\\??=\\|\\??<<\\|\\??>>\\|\\??<\\|\\??>\\|:=\\|\\??<=\\|\\??>=\\|=>\\|\\??/=\\|\\?\\?\\)\\([^=>\n]\\|$\\)\\)" end t)
        //       (if (or (match-string 1)
        // 	      (<= (match-beginning 0)  ; not if at boi
        // 		 (save-excursion (back-to-indentation) (point))))
        // 	  (goto-char (match-end 0))
        // 	(replace-match "\\3 \\4 \\5")
        // 	(goto-char (match-end 2))))
        //     ;; eliminate multiple spaces and spaces at end of line
        //     (goto-char beg)
        //     (while (or (and (looking-at "--.*\n") (re-search-forward "--.*\n" end t))
        // 	       (and (looking-at "--.*") (re-search-forward "--.*" end t))
        // 	       (and (looking-at "\"") (re-search-forward "\"[^\"\n]*[\"\n]" end t))
        // 	       (and (looking-at "\\s-+$") (re-search-forward "\\s-+$" end t)
        // 		    (progn (replace-match "" nil nil) t))
        // 	       (and (looking-at "\\s-+;") (re-search-forward "\\s-+;" end t)
        // 		    (progn (replace-match ";" nil nil) t))
        // 	       (and (looking-at "^\\s-+") (re-search-forward "^\\s-+" end t))
        // 	       (and (looking-at "\\s-+--") (re-search-forward "\\s-+" end t)
        // 		    (progn (replace-match "  " nil nil) t))
        // 	       (and (looking-at "\\s-+") (re-search-forward "\\s-+" end t)
        // 		    (progn (replace-match " " nil nil) t))
        // 	       (and (looking-at "-") (re-search-forward "-" end t))
        // 	       (re-search-forward "[^ \t\"-]+" end t))))
        //   (unless no-message (message "Fixing up whitespace...done")))
    }
}

const test = `asdasd,asdasd
asdasdasd ; 
asdasd`;
const formatter = new Formatter(test);
formatter.vhdlBeautifyRegion();
console.log(formatter.text);