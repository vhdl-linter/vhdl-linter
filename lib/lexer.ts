import { OFile, OIRange, ParserError } from './parser/objects';

export enum TokenType {
  basicIdentifier = 'basicIdentifier',
  extendedIdentifier = 'extendedIdentifier',
  decimalLiteral = 'decimalLiteral',
  characterLiteral = 'characterLiteral',
  stringLiteral = 'stringLiteral',
  bitStringLiteral = 'bitStringLiteral',
  whitespace = 'whitespace',
  comment = 'comment',
  delimiter = 'delimiter',
  keyword = 'keyword',
}

export class OLexerToken {
  constructor(
    public text: string,
    public range: OIRange,
    public type: TokenType,
    public file: OFile
  ) {
  }
  isIdentifier() {
    return this.type === TokenType.basicIdentifier || this.type === TokenType.extendedIdentifier;
  }
  isDesignator() {
    return this.type === TokenType.basicIdentifier || this.type === TokenType.extendedIdentifier
      || this.type === TokenType.stringLiteral || this.type === TokenType.characterLiteral;
  }
  isLiteral() {
    return this.type === TokenType.decimalLiteral || this.type === TokenType.bitStringLiteral
      || this.type === TokenType.stringLiteral || this.type === TokenType.characterLiteral || this.getLText() === 'null';

  }
  isWhitespace() {
    return this.type === TokenType.whitespace || this.type === TokenType.comment;
  }
  getLText() {
    return this.text.toLowerCase();
  }
  toString() {
    return this.text;
  }
}
export class Lexer {
  readonly reservedWords = new RegExp('^(' + [
    'abs', 'access', 'after', 'alias', 'all', 'and', 'architecture', 'array', 'assert', 'assume', 'assume_guarantee', 'attribute', 'begin', 'block', 'body', 'buffer',
    'bus', 'case', 'component', 'configuration', 'constant', 'context', 'cover', 'default', 'disconnect', 'downto', 'else', 'elsif', 'end', 'entity', 'exit', 'fairness',
    'file', 'for', 'force', 'function', 'generate', 'generic', 'group', 'guarded', 'if', 'impure', 'in', 'inertial', 'inout', 'is', 'label', 'library',
    'linkage', 'literal', 'loop', 'map', 'mod', 'nand', 'new', 'next', 'nor', 'not', 'null', 'of', 'on', 'open', 'or', 'others',
    'out', 'package', 'parameter', 'port', 'postponed', 'procedure', 'process', 'property', 'protected', 'pure', 'range', 'record', 'register', 'reject', 'release', 'rem',
    'report', 'restrict', 'restrict_guarantee', 'return', 'rol', 'ror', 'select', 'sequence', 'severity', 'signal', 'shared', 'sla', 'sll', 'sra', 'srl', 'strong',
    'subtype', 'then', 'to', 'transport', 'type', 'unaffected', 'units', 'until', 'use', 'variable', 'vmode', 'vprop', 'vunit', 'wait', 'when', 'while', 'with', 'xnor', 'xor',
  ].join('|') + ')\\b', 'i');

  tokenTypes: { regex: RegExp, tokenType: TokenType }[] = [
    { regex: /^--.*/, tokenType: TokenType.comment },
    { regex: /^\s+/i, tokenType: TokenType.whitespace },
    { regex: /^[0-9]+#[0-9a-z][0-9_a-z]*(?:\.[0-9a-z][0-9_a-z]+)?#(?:e[+-]?[0-9_]+)?/i, tokenType: TokenType.decimalLiteral },
    { regex: /^-?[0-9_]+(?:\.[0-9_]+)?(?:e[+-]?[0-9_]+)?/i, tokenType: TokenType.decimalLiteral },
    { regex: /^[0-9]*(?:B|O|X|UB|UO|UX|SB|SO|SX|D)"[^"]+"/i, tokenType: TokenType.bitStringLiteral },
    { regex: /^[a-z][a-z0-9_]*/i, tokenType: TokenType.basicIdentifier },
    { regex: /^\\.*?[^\\]\\(?!\\)/i, tokenType: TokenType.extendedIdentifier },
    // BASED LITERAL
    { regex: /^'.'/, tokenType: TokenType.characterLiteral },
    { regex: /^"(?:[^"]|(?:""))*"(?!")/i, tokenType: TokenType.stringLiteral },
    {
      regex: /^=>|\*\*|:=|\/=>=|<=|<>|\?\?|\?=|\?\/=|\/=|\?<|\?<=|\?>|\?>=|<<|>>|&|'|\(|\)|\*|\+|,|-|.|\/|:|;|<|=|>|`|\||\[|\]|\?|@/,
      tokenType: TokenType.delimiter
    },

  ];
  constructor(
    public text: string,
    public file: OFile
  ) { }
  lex(file: OFile) {
    const tokens: OLexerToken[] = [];
    let foundToken = true;
    let offset = 0;
    let lastOffset = -1;
    let text = this.text;
    loop: while (text.length > 0 && foundToken) {
      // if (iterationCount++ > 10000) {
      //   throw new ParserError('Lexer infinite loop2!', new OIRange(this.file, offset, offset));
      // }
      if (lastOffset === offset) {
        throw new ParserError('Lexer infinite loop!', new OIRange(this.file, offset, offset));
      }
      lastOffset = offset;
      foundToken = false;
      const match = text.match(this.reservedWords);
      if (match) {
        const token = new OLexerToken(match[0],
          new OIRange(this.file, offset, offset + match[0].length), TokenType.keyword, file);

        tokens.push(token);
        text = text.substring(match[0].length);
        offset += match[0].length;
        foundToken = true;

        continue loop;
      }
      for (const tokenType of this.tokenTypes) {
        const match = text.match(tokenType.regex);
        if (match) {
          const token = new OLexerToken(match[2] ? match[2] : match[0],
            new OIRange(this.file, offset, offset + match[0].length), tokenType.tokenType, file);

          tokens.push(token);
          text = text.substring(match[0].length);
          offset += match[0].length;
          foundToken = true;
          continue loop;
        }
      }
      if (!foundToken) {
        throw new ParserError('Lexer stuck!', new OIRange(this.file, offset, offset));
      }
    }
    // console.log(tokens);
    return tokens;
  }
}