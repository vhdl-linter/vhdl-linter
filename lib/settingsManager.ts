import { connection, documents, validateTextDocument } from "./languageServer";
import { defaultSettings, ISettings, normalizeSettings } from "./settings";

let globalSettings: ISettings = defaultSettings;
export const currentCapabilities = {
  configuration: false,
  workspaceFolder: false
};
// Cache the settings of all open documents
export const documentSettings = new Map<string, ISettings>();

export async function getDocumentSettings(resource?: URL): Promise<ISettings> {
  // default settings are assumed as default and the overwritten by either
  // settings from vs code (workspace) or the closest vhdl-linter.yml
  if (!currentCapabilities.configuration) {
    return normalizeSettings(globalSettings);
  }
  let result = documentSettings.get(resource?.toString() ?? '');
  if (!result) {
    result = normalizeSettings(await connection.workspace.getConfiguration({
      scopeUri: resource?.toString(),
      section: 'VhdlLinter'
    }) as ISettings);
  }
  documentSettings.set(resource?.toString() ?? '', result);
  return result;
}

export function changeConfigurationHandler(change: { settings: { VhdlLinter?: ISettings } }) {

  if (currentCapabilities.configuration) {
    // Reset all cached document settings
    documentSettings.clear();
  } else {
    globalSettings = (
      (change.settings.VhdlLinter ?? defaultSettings)
    );
  }

  // Revalidate all open text documents
  for (const document of documents.all()) {
    void validateTextDocument(document);
  }
}