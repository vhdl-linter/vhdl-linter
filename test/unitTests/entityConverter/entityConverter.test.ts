import { afterAll, beforeAll, describe, expect, test } from "@jest/globals";
import { join } from "path";
import { pathToFileURL } from "url";
import { Position } from "vscode-languageserver";
import { converterTypes, entityConverter } from "../../../lib/entityConverter";
import { ProjectParser } from "../../../lib/projectParser";
import { VhdlLinter } from "../../../lib/vhdlLinter";
import { readFileSyncNorm } from "../../readFileSyncNorm";
import { getDocumentSettings, overwriteSettings } from "../../../lib/settingsManager";

let projectParser: ProjectParser;
beforeAll(async () => {
  projectParser = await ProjectParser.create([pathToFileURL(__dirname)]);
});
afterAll(async () => {
  await projectParser.stop();
});
describe('Testing entityConverter', () => {
  test('Testing entity converter for with different configurations', async () => {
    const path = join(__dirname, 'test_entity.vhd');
    const uri = pathToFileURL(path);
    const linter = new VhdlLinter(uri, readFileSyncNorm(uri, { encoding: 'utf8' }), projectParser, await getDocumentSettings(uri, projectParser));
    for (const type of ['instance', 'signals', 'sysverilog', 'component'] as converterTypes[]) {
      const overwritesStyles = {
        instantiationLabelPrefix: ['', 'instPrefix_'],
        instantiationLabelSuffix: ['_instSuffix'],
        signalPrefix: ['sigPrefix_'],
        signalSuffix: ['_sigSuffix'],
        genericPrefix: ['genericPrefix_'],
        genericSuffix: ['_genericSuffix'],
      };
      for (const key of Object.keys(overwritesStyles) as (keyof typeof overwritesStyles)[]) {
        for (const value of overwritesStyles[key]) {
          const overwrite = { style: { [key]: value } };
          const settings = overwriteSettings(await getDocumentSettings(pathToFileURL(path), projectParser), overwrite);
          const template = entityConverter(linter, type, settings);
          expect(template).toMatchSnapshot(`type ${type} overwrite ${JSON.stringify(overwrite)}`);

        }
      }
    }
  });
  test('Testing position based fetch of correct entity', async () => {
    const path = join(__dirname, 'two_entities.vhd');
    const uri = pathToFileURL(path);
    const linter = new VhdlLinter(uri, readFileSyncNorm(uri, { encoding: 'utf8' }), projectParser, await getDocumentSettings(uri, projectParser));
    {
      const template = entityConverter(linter, 'instance', await getDocumentSettings(uri, projectParser));
      expect(template).toContain('first_entity');
    }
    {
      const template = entityConverter(linter, 'instance', await getDocumentSettings(uri, projectParser), Position.create(10, 10));
      expect(template).toContain('first_entity');
    }
    {
      const template = entityConverter(linter, 'instance', await getDocumentSettings(uri, projectParser), Position.create(16, 5));
      expect(template).toContain('second_entity');
    }
    {
      const template = entityConverter(linter, 'instance', await getDocumentSettings(uri, projectParser), Position.create(25, 0));
      expect(template).toContain('first_entity');
    }
  });
  test('Testing fetching in empty file', async () => {
    const path = join(__dirname, 'fileWithoutEntity.vhd');
    const uri = pathToFileURL(path);
    const linter = new VhdlLinter(uri, readFileSyncNorm(uri, { encoding: 'utf8' }), projectParser, await getDocumentSettings(uri, projectParser));
    {
      const template = entityConverter(linter, 'instance', await getDocumentSettings(uri, projectParser));
      expect(template).toBeUndefined();
    }
    {
      const template = entityConverter(linter, 'instance', await getDocumentSettings(uri, projectParser), Position.create(25, 0));
      expect(template).toBeUndefined();
    }
    {
      const template = entityConverter(linter, 'instance', await getDocumentSettings(uri, projectParser), Position.create(31, 0));
      expect(template).toBeUndefined();
    }

  });
});