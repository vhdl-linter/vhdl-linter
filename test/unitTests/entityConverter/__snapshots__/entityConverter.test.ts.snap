// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing entityConverter Testing entity converter for with different configurations: type instance overwrite {"style":{"genericPrefix":"genericPrefix_"}} 1`] = `
"inst_test_entity : entity work.test_entity
generic map (
  LENGTH => genericPrefix_LENGTH
)
port map (
  a => a,
  b => b,
  c => c
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type instance overwrite {"style":{"genericSuffix":"_genericSuffix"}} 1`] = `
"inst_test_entity : entity work.test_entity
generic map (
  LENGTH => LENGTH_genericSuffix
)
port map (
  a => a,
  b => b,
  c => c
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type instance overwrite {"style":{"instantiationLabelPrefix":""}} 1`] = `
"test_entity : entity work.test_entity
generic map (
  LENGTH => LENGTH
)
port map (
  a => a,
  b => b,
  c => c
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type instance overwrite {"style":{"instantiationLabelPrefix":"instPrefix_"}} 1`] = `
"instPrefix_test_entity : entity work.test_entity
generic map (
  LENGTH => LENGTH
)
port map (
  a => a,
  b => b,
  c => c
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type instance overwrite {"style":{"instantiationLabelSuffix":"_instSuffix"}} 1`] = `
"inst_test_entity_instSuffix : entity work.test_entity
generic map (
  LENGTH => LENGTH
)
port map (
  a => a,
  b => b,
  c => c
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type instance overwrite {"style":{"signalPrefix":"sigPrefix_"}} 1`] = `
"inst_test_entity : entity work.test_entity
generic map (
  LENGTH => LENGTH
)
port map (
  a => sigPrefix_a,
  b => sigPrefix_b,
  c => sigPrefix_c
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type instance overwrite {"style":{"signalSuffix":"_sigSuffix"}} 1`] = `
"inst_test_entity : entity work.test_entity
generic map (
  LENGTH => LENGTH
)
port map (
  a => a_sigSuffix,
  b => b_sigSuffix,
  c => c_sigSuffix
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type signals overwrite {"style":{"genericPrefix":"genericPrefix_"}} 1`] = `
"signal a : std_ulogic;
signal b : std_ulogic_vector(7 downto 0);
signal c : std_ulogic_vector(LENGTH -1 downto 0);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type signals overwrite {"style":{"genericSuffix":"_genericSuffix"}} 1`] = `
"signal a : std_ulogic;
signal b : std_ulogic_vector(7 downto 0);
signal c : std_ulogic_vector(LENGTH -1 downto 0);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type signals overwrite {"style":{"instantiationLabelPrefix":""}} 1`] = `
"signal a : std_ulogic;
signal b : std_ulogic_vector(7 downto 0);
signal c : std_ulogic_vector(LENGTH -1 downto 0);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type signals overwrite {"style":{"instantiationLabelPrefix":"instPrefix_"}} 1`] = `
"signal a : std_ulogic;
signal b : std_ulogic_vector(7 downto 0);
signal c : std_ulogic_vector(LENGTH -1 downto 0);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type signals overwrite {"style":{"instantiationLabelSuffix":"_instSuffix"}} 1`] = `
"signal a : std_ulogic;
signal b : std_ulogic_vector(7 downto 0);
signal c : std_ulogic_vector(LENGTH -1 downto 0);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type signals overwrite {"style":{"signalPrefix":"sigPrefix_"}} 1`] = `
"signal sigPrefix_a : std_ulogic;
signal sigPrefix_b : std_ulogic_vector(7 downto 0);
signal sigPrefix_c : std_ulogic_vector(LENGTH -1 downto 0);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type signals overwrite {"style":{"signalSuffix":"_sigSuffix"}} 1`] = `
"signal a_sigSuffix : std_ulogic;
signal b_sigSuffix : std_ulogic_vector(7 downto 0);
signal c_sigSuffix : std_ulogic_vector(LENGTH -1 downto 0);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type sysverilog overwrite {"style":{"genericPrefix":"genericPrefix_"}} 1`] = `
"test_entity #(
  .LENGTH(genericPrefix_LENGTH)
) inst_test_entity (
  .a(a),
  .b(b),
  .c(c)
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type sysverilog overwrite {"style":{"genericSuffix":"_genericSuffix"}} 1`] = `
"test_entity #(
  .LENGTH(LENGTH_genericSuffix)
) inst_test_entity (
  .a(a),
  .b(b),
  .c(c)
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type sysverilog overwrite {"style":{"instantiationLabelPrefix":""}} 1`] = `
"test_entity #(
  .LENGTH(LENGTH)
) test_entity (
  .a(a),
  .b(b),
  .c(c)
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type sysverilog overwrite {"style":{"instantiationLabelPrefix":"instPrefix_"}} 1`] = `
"test_entity #(
  .LENGTH(LENGTH)
) instPrefix_test_entity (
  .a(a),
  .b(b),
  .c(c)
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type sysverilog overwrite {"style":{"instantiationLabelSuffix":"_instSuffix"}} 1`] = `
"test_entity #(
  .LENGTH(LENGTH)
) inst_test_entity_instSuffix (
  .a(a),
  .b(b),
  .c(c)
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type sysverilog overwrite {"style":{"signalPrefix":"sigPrefix_"}} 1`] = `
"test_entity #(
  .LENGTH(LENGTH)
) inst_test_entity (
  .a(sigPrefix_a),
  .b(sigPrefix_b),
  .c(sigPrefix_c)
);
"
`;

exports[`Testing entityConverter Testing entity converter for with different configurations: type sysverilog overwrite {"style":{"signalSuffix":"_sigSuffix"}} 1`] = `
"test_entity #(
  .LENGTH(LENGTH)
) inst_test_entity (
  .a(a_sigSuffix),
  .b(b_sigSuffix),
  .c(c_sigSuffix)
);
"
`;
